// powermax-api/lib/tables.js
"use strict";

const PowerMaxTables = {
	// PowerMax messages we send
	// E1: message string
	// E2: expected response after ACK (null if none)
	send: {
		MSG_INIT    : { msg: [0xAB, 0x0A, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x43] },
		MSG_ZONENAME: { msg: [0xA3, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x43] },
		MSG_ZONETYPE: { msg: [0xA6, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x43] },
		MSG_X10NAMES: { msg: [0xAC, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x43] },
		MSG_RESTORE : { msg: [0xAB, 0x06, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x43], resp: 0xA5 },
		MSG_ENROLL  : { msg: [0xAB, 0x0A, 0x00, 0x00, 'code:2', 0x00, 0x00, 0x00, 0x00, 0x00, 0x43] },
		MSG_EVENTLOG: { msg: [0xA0, 0x00, 0x00, 0x00, 'pin:2', 0x00, 0x00, 0x00, 0x00, 0x00, 0x43], resp: 0xA0 },
		MSG_ARM     : { msg: [0xA1, 0x00, 0x00, 'arm', 'pin:2', 0x00, 0x00, 0x00, 0x00, 0x00, 0x43] },
		MSG_STATUS  : { msg: [0xA2, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x43], resp: 0xA5 },
		MSG_BYPASSTA: { msg: [0xA2, 0x00, 0x00, 0x20, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x43], resp: 0xA5 },
		MSG_X10PGM  : { msg: [0xA4, 0x00, 0x00, 0x00, 0x00, 0x00, 'cmd', 'device:2', 0x00, 0x00, 0x43] },
		MSG_BYPASSEN: { msg: [0xAA, 'pin:2', 'bypass:4', 0x00, 0x00, 0x00, 0x00, 0x43] },
		MSG_BYPASSDI: { msg: [0xAA, 'pin:2', 0x00, 0x00, 0x00, 0x00, 'bypass:4', 0x43] },
		MSG_X10NAME : { msg: [0xAC, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x43], resp: 0xAC },
		MSG_DOWNLOAD: { msg: [0x24, 0x00, 0x00, 'code:2', 0x00, 0x00, 0x00, 0x00, 0x00, 0x00], resp: 0x3C },
		MSG_SETTIME : { msg: [0x46, 0xF8, 0x00, 'time:6', 0xFF, 0xFF] },
		MSG_WRITE   : { msg: [0x3D, 'addr', 'len', 'val'] },
		MSG_DL      : { msg: [0x3E, 'item:4', 0xB0, 0x00, 0x00, 0x00, 0x00, 0x00], resp: 0x3F },
		MSG_SER_TYPE: { msg: [0x5A, 0x30, 0x04, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00], resp: 0x33 },
		MSG_START   : { msg: [0x0A], resp: 0x33 },
		MSG_EXIT    : { msg: [0x0F] },
		// PowerMaster specific
		MSG_POWERMASTER : { msg: [0xB0, 0x01, 'msg', 0x43], resp: 0xB0 }
	},

	// Download messages
	download: {
		MSG_DL_TIME      : [0xF8, 0x00, 0x06, 0x00],
		MSG_DL_COMMDEF   : [0x01, 0x01, 0x1E, 0x00],
		MSG_DL_PHONENRS  : [0x36, 0x01, 0x20, 0x00],
		MSG_DL_PINCODES  : [0xFA, 0x01, 0x10, 0x00],
		MSG_DL_PGMX10    : [0x14, 0x02, 0xD5, 0x00],
		MSG_DL_PARTITIONS: [0x00, 0x03, 0xF0, 0x00],
		MSG_DL_PANELFW   : [0x00, 0x04, 0x20, 0x00],
		MSG_DL_SERIAL    : [0x30, 0x04, 0x08, 0x00],
		MSG_DL_ZONES     : [0x00, 0x09, 0x78, 0x00],
		MSG_DL_KEYFOBS   : [0x78, 0x09, 0x40, 0x00],
		MSG_DL_2WKEYPAD  : [0x00, 0x0A, 0x08, 0x00],
		MSG_DL_1WKEYPAD  : [0x20, 0x0A, 0x40, 0x00],
		MSG_DL_SIRENS    : [0x60, 0x0A, 0x08, 0x00],
		MSG_DL_X10NAMES  : [0x30, 0x0B, 0x10, 0x00],
		MSG_DL_ZONENAMES : [0x40, 0x0B, 0x1E, 0x00],
		MSG_DL_EVENTLOG  : [0xDF, 0x04, 0x28, 0x03],
		MSG_DL_ZONESTR   : [0x00, 0x19, 0x00, 0x02],
		MSG_DL_ZONECUSTOM: [0xA0, 0x1A, 0x50, 0x00],
		MSG_DL_MR_ZONENAMES: [0x60, 0x09, 0x40, 0x00],
		MSG_DL_MR_PINCODES : [0x98, 0x0A, 0x60, 0x00],
		MSG_DL_MR_SIRENS   : [0xE2, 0xB6, 0x50, 0x00],
		MSG_DL_MR_KEYPADS  : [0x32, 0xB7, 0x40, 0x01],
		MSG_DL_MR_ZONES    : [0x72, 0xB8, 0x80, 0x02],
		MSG_DL_MR_SIRKEYZON: [0xE2, 0xB6, 0x10, 0x04], // Combines Sirens, keypads and sensors
		MSG_DL_ALL       : [0x00, 0x00, 0x00, 0xFF]
	},

	// Message types we can receive with their length (null=variable), handlers and whether they need an ACK
	receive: {
		'0x02': { handler: 'handleAck', ack: false },
		'0x06': { handler: 'handleTimeout', ack: false },
		'0x08': { handler: 'handleDenied', ack: true },
		'0x0B': { handler: 'handleStop', ack: true },
		'0x25': { len: 14, handler: 'handleDownloadRetry', ack: true },
		'0x33': { len: 14, handler: 'handleSettings', ack: true },
		'0x3C': { len: 14, handler: 'handleInfo', ack: true },
		'0x3F': { handler: 'handleSettingsItem', ack: true },
		'0xA0': { len: 15, handler: 'handleEventlog', ack: true },
		'0xA3': { len: 15, handler: 'handleZoneName', ack: true },
		'0xA5': { len: 15, handler: 'handleStatus', ack: true },
		'0xA6': { len: 15, handler: 'handleZoneType', ack: true },
		'0xA7': { len: 15, handler: 'handlePanel', ack: true },
		'0xAB': { len: 15, handler: 'handlePowerlink', ack: false },
		'0xB0': { handler: 'handlePowerMaster', ack: true },
		'0xF1': { len: 9, ack: false }
	},

	// PowerMaster messages
	receiveB0: {
		'0x04': { txt: "Zone status" },
		'0x18': { txt: "Open/close status" },
		'0x39': { txt: "Activity" }
	},

	// Log event entries
	logEvent: {
		en: [
			"None", "Interior Alarm", "Perimeter Alarm", "Delay Alarm", "24h Silent Alarm", "24h Audible Alarm",
			"Tamper", "Control Panel Tamper", "Tamper Alarm", "Tamper Alarm", "Communication Loss", "Panic From Keyfob",
			"Panic From Control Panel", "Duress", "Confirm Alarm", "General Trouble", "General Trouble Restore",
			"Interior Restore", "Perimeter Restore", "Delay Restore", "24h Silent Restore", "24h Audible Restore",
			"Tamper Restore", "Control Panel Tamper Restore", "Tamper Restore", "Tamper Restore", "Communication Restore",
			"Cancel Alarm", "General Restore", "Trouble Restore", "Not used", "Recent Close", "Fire", "Fire Restore",
			"No Active", "Emergency", "No used", "Disarm Latchkey", "Panic Restore", "Supervision (Inactive)",
			"Supervision Restore (Active)", "Low Battery", "Low Battery Restore", "AC Fail", "AC Restore",
			"Control Panel Low Battery", "Control Panel Low Battery Restore", "RF Jamming", "RF Jamming Restore",
			"Communications Failure", "Communications Restore", "Telephone Line Failure", "Telephone Line Restore",
			"Auto Test", "Fuse Failure", "Fuse Restore", "Keyfob Low Battery", "Keyfob Low Battery Restore", "Engineer Reset",
			"Battery Disconnect", "1-Way Keypad Low Battery", "1-Way Keypad Low Battery Restore", "1-Way Keypad Inactive",
			"1-Way Keypad Restore Active", "Low Battery", "Clean Me", "Fire Trouble", "Low Battery", "Battery Restore",
			"AC Fail", "AC Restore", "Supervision (Inactive)", "Supervision Restore (Active)", "Gas Alert", "Gas Alert Restore",
			"Gas Trouble", "Gas Trouble Restore", "Flood Alert", "Flood Alert Restore", "X-10 Trouble", "X-10 Trouble Restore",
			"Arm Home", "Arm Away", "Quick Arm Home", "Quick Arm Away", "Disarm", "Fail To Auto-Arm", "Enter To Test Mode",
			"Exit From Test Mode", "Force Arm", "Auto Arm", "Instant Arm", "Bypass", "Fail To Arm", "Door Open",
			"Communication Established By Control Panel", "System Reset", "Installer Programming", "Wrong Password",
			"Not Sys Event", "Not Sys Event", "Extreme Hot Alert", "Extreme Hot Alert Restore", "Freeze Alert",
			"Freeze Alert Restore", "Human Cold Alert", "Human Cold Alert Restore", "Human Hot Alert",
			"Human Hot Alert Restore", "Temperature Sensor Trouble", "Temperature Sensor Trouble Restore",
			// new values partition models
			"PIR Mask", "PIR Mask Restore", "", "", "", "", "", "", "", "", "", "",
			"Alarmed", "Restore", "Alarmed", "Restore", "", "", "", "", "", "", "", "", "", "",
			"", "", "", "", "", "Exit Installer", "Enter Installer", "", "", "", "", ""
		],
		nl: [
			"Geen", "In alarm", "In alarm", "In alarm", "In alarm", "In alarm",
			"Sabotage alarm", "Systeem sabotage", "Sabotage alarm", "Add user", "Communicate fout", "Paniekalarm",
			"Code bedieningspaneel paniek", "Dwang", "Bevestig alarm", "Successful U/L", "Probleem herstel",
			"Herstel", "Herstel", "Herstel", "Herstel", "Herstel",
			"Sabotage herstel", "Systeem sabotage herstel", "Sabotage herstel", "Sabotage herstel", "Communicatie herstel",
			"Stop alarm", "Algemeen herstel", "Brand probleem herstel", "Systeem inactief", "Recent close", "Brand", "Brand herstel",
			"Niet actief", "Noodoproep", "Remove user", "Controleer code", "Bevestig alarm", "Supervisie",
			"Supervisie herstel", "Batterij laag", "Batterij OK", "230VAC uitval", "230VAC herstel",
			"Controlepaneel batterij laag", "Controlepaneel batterij OK", "Radio jamming", "Radio herstel",
			"Communicatie mislukt", "Communicatie hersteld", "Telefoonlijn fout", "Telefoonlijn herstel",
			"Automatische test", "Zekeringsfout", "Zekering herstel", "Batterij laag", "Batterij OK", "Monteur reset",
			"Accu vermist", "Batterij laag", "Batterij OK", "Supervisie",
			"Supervisie herstel", "Lage batterij bevestiging", "Reinigen", "Probleem", "Batterij laag", "Batterij OK",
			"230VAC uitval", "230VAC herstel", "Supervisie", "Supervisie herstel", "Gas alarm", "Gas herstel",
			"Gas probleem", "Gas probleem herstel", "Lekkage alarm", "Lekkage herstel", "Probleem", "Probleem herstel",
			"Deelschakeling", "Ingeschakeld", "Snel deelschakeling", "Snel ingeschakeld", "Uitgezet", "Inschakelfout (auto)", "Test gestart",
			"Test gestopt", "Force aan", "Geheel in (auto)", "Onmiddelijk", "Overbruggen", "Inschakelfout",
			"Log verzenden", "Systeem reset", "Installateur programmeert", "Foutieve code", "Overbruggen"
		]
	},

	// Log user values
	logUser: {
		en: [
			"System", "Zone 1", "Zone 2", "Zone 3", "Zone 4", "Zone 5", "Zone 6", "Zone 7", "Zone 8",
			"Zone 9", "Zone 10", "Zone 11", "Zone 12", "Zone 13", "Zone 14", "Zone 15", "Zone 16", "Zone 17", "Zone 18",
			"Zone 19", "Zone 20", "Zone 21", "Zone 22", "Zone 23", "Zone 24", "Zone 25", "Zone 26", "Zone 27", "Zone 28",
			"Zone 29", "Zone 30", "Keyfob 1", "Keyfob 2", "Keyfob 3", "Keyfob 4", "Keyfob 5", "Keyfob 6", "Keyfob 7", "Keyfob 8",
			"User 1", "User 2", "User 3", "User 4", "User 5", "User 6", "User 7", "User 8", "Pad 1", "Pad 2",
			"Pad  3", "Pad  4", "Pad 5", "Pad  6", "Pad  7", "Pad 8", "Sirine 1", "Sirene 2", "2Pad 1", "2Pad 2",
			"2Pad 3", "2Pad 4", "X10-1", "X10-2", "X10-3", "X10-4", "X10-5", "X10-6", "X10-7", "X10-8",
			"X10-9", "X10-10", "X10-11", "X10-12", "X10-13", "X10-14", "X10-15", "PGM", "GSM", "Powerlink",
			"Proximity tag 1", "Proximity tag 2", "Proximity tag 3", "Proximity tag 4", "Proximity tag 5", "Proximity tag 6", "Proximity tag 7", "Proximity tag 8"
		],
		nl: [
			"Systeem", "Zone 1", "Zone 2", "Zone 3", "Zone 4", "Zone 5", "Zone 6", "Zone 7", "Zone 8",
			"Zone 9", "Zone 10", "Zone 11", "Zone 12", "Zone 13", "Zone 14", "Zone 15", "Zone 16", "Zone 17", "Zone 18",
			"Zone 19", "Zone 20", "Zone 21", "Zone 22", "Zone 23", "Zone 24", "Zone 25", "Zone 26", "Zone 27", "Zone 28",
			"Zone 29", "Zone 30", "Keyfob 1", "Keyfob 2", "Keyfob 3", "Keyfob 4", "Keyfob 5", "Keyfob 6", "Keyfob 7", "Keyfob 8",
			"Gebruiker 1", "Gebruiker 2", "Gebruiker 3", "Gebruiker 4", "Gebruiker 5", "Gebruiker 6", "Gebruiker 7", "Gebruiker 8", "Pad  01", "Pad  02",
			"Pad 3", "Pad  4", "Pad 5", "Pad 6", "Pad 7", "Pad 8", "Sirene 1", "Sirine 2", "2Pad 1", "2Pad 2",
			"2Pad 3", "2Pad 4", "X10-1", "X10-2", "X10-3", "X10-4", "X10-5", "X10-6", "X10-7", "X10-8",
			"X10-9", "X10-10", "X10-11", "X10-12", "X10-13", "X10-14", "X10-15", "PGM", "GSM", "Powerlink",
			"Proximity tag 1", "Proximity tag 2", "Proximity tag 3", "Proximity tag 4", "Proximity tag 5", "Proximity tag 6", "Proximity tag 7", "Proximity tag 8"
		]
	},

	// System status
	sysStatus: {
		en: [
			"Disarmed", "Home Exit Delay", "Away Exit Delay", "Entry Delay", "Armed Home", "Armed Away", "User Test",
			"Downloading", "Programming", "Installer", "Home Bypass", "Away Bypass", "Ready", "Not Ready", "??", "??",
			"Disarmed Instant", "Home Instant Exit Delay", "Away Instant Exit Delay", "Entry Delay Instant", "Armed Home Instant",
			"Armed Away Instant"
		],
		nl: [
			"Uitgeschakeld", "Deel uitloopvertraging", "Totaal uitloopvertraging", "Inloopvertraging", "Deel ingeschakeld",
			"Totaal ingeschakeld", "Gebruiker test", "Downloaden", "Programmeren", "Monteurmode", "Deel met overbrugging",
			"Totaal met overbrugging", "Klaar", "Niet klaar", "??", "??", "Direct uitschakelen", "Direct Deel uitloopvertraging",
			"Direct Totaal uitloopvertraging", "Direct inloopvertraging", "Direct Deel", "Direct Totaal"
		]
	},

	// System status flags
	sysStatusFlags: {
		en: [
			"Ready", "Alert in memory", "Trouble", "Bypass on", "Last 10 seconds", "Zone event", "Status changed", "Alarm event"
		],
		nl: [
			"Klaar", "Alarm in geheugen", "Probleem", "Overbruggen aan", "Laatste 10 seconden", "Zone verstoord", "Status gewijzigd",
			"Alarm actief"
		]
	},

	// isArmed lookup table
	isArmed: {
		0x03: true, 0x04: true, 0x05: true, 0x0A: true, 0x0B: true, 0x13: true, 0x14: true, 0x15: true
	},

	// Arming modes
	armMode: {
		Disarmed: 0x00, Stay: 0x04, Armed: 0x05, UserTest: 0x06, StayInstant: 0x14, ArmedInstant: 0x15,
		Night: 0x04, NightInstant: 0x14
	},

	// Detailed arming modes (Not used: Night, NightInstant, Vacation)
	detailedArmMode: [
		"Disarmed", "ExitDelay", "ExitDelay", "EntryDelay", "Stay", "Armed", "UserTest", "NotReady", "NotReady", "NotReady",
		"Force", "Force", "Ready", "NotReady", "??", "??", "Disarmed", "ExitDelay", "ExitDelay", "EntryDelay", "StayInstant",
		"ArmedInstant"
	],

	// Event types
	eventType: {
		en: [
			"None", "Tamper Alarm", "Tamper Restore", "Open", "Closed", "Violated (Motion)", "Panic Alarm", "RF Jamming",
			"Tamper Open", "Communication Failure", "Line Failure", "Fuse", "Not Active", "Low Battery", "AC Failure",
			"Fire Alarm", "Emergency", "Siren Tamper", "Siren Tamper Restore", "Siren Low Battery", "Siren AC Fail"
		],
		nl: [
			"Geen", "Sabotage alarm", "Sabotage herstel", "Open", "Gesloten", "Verstoord (beweging)", "Paniek alarm", "RF verstoring",
			"Sabotage open", "Communicatie probleem", "Lijnfout", "Zekering", "Niet actief", "Lage batterij", "AC probleem",
			"Brandalarm", "Noodoproep", "Sirene sabotage", "Sirene sabotage herstel", "Sirene lage batterij", "Sirene AC probleem"
		]
	},

	// Alarm type of the panel
	panelAlarmType: {
		'0x01': "Intruder", '0x02': "Intruder", '0x03': "Intruder", '0x04': "Intruder", '0x05': "Intruder", '0x06': "Tamper",
		'0x07': "Tamper", '0x08': "Tamper", '0x09': "Tamper", '0x0b': "Panic", '0x0c': "Panic", '0x20': "Fire",
		'0x23': "Emergency", '0x49': "Gas", '0x4c': "Flood"
	},

	// Translations for alarm types
	panelAlarmText: {
		en: { Intruder: "Intruder", Tamper: "Tamper", Panic: "Panic", Fire: "Fire", Emergency: "Emergency", Gas: "Gas", Flood: "Flood" },
		nl: { Intruder: "Indringer", Tamper: "Knoeien", Panic: "Paniek", Fire: "Brand", Emergency: "Nood", Gas: "Gas", Flood: "Overstroming" }
	},

	// The kind of trouble
	panelTroubleType: {
		'0x0a': "Communication", '0x0f': "General", '0x29': "Battery", '0x2c': "Power", '0x2d': "Battery", '0x2f': "Jamming",
		'0x31': "Communication", '0x33': "Telephone", '0x36': "Power", '0x38': "Battery", '0x3b': "Battery", '0x3c': "Battery",
		'0x40': "Battery", '0x43': "Battery"
	},

	// Translations for trouble types
	panelTroubleText: {
		en: { Communication: "Communication", General: "General", Battery: "Battery", Power: "Power", Jamming: "Jamming", Telephone: "Telephone" },
		nl: { Communication: "Communicatie", General: "Algemeen", Battery: "Batterij", Power: "Stroom", Jamming: "Storen", Telephone: "Telefoon" }
	},

	// Type of panel
	panelType: {
		'1': "PowerMax", '2': "PowerMax+", '3': "PowerMax Pro", '4': "PowerMax Complete", '5': "PowerMax Pro Part",
		'6': "PowerMax Complete Part", '7': "PowerMax Express", '8': "PowerMaster10", '9': "PowerMaster30"
	},

	// Config for each panel type (1-9)
	panelConfig: {
		CFG_PARTITIONS: [   1,   1,   1,   1,   3,   3,   1,   3,   3 ],
		CFG_EVENTS    : [ 250, 250, 250, 250, 250, 250, 250, 250,1000 ],
		CFG_KEYFOBS   : [   8,   8,   8,   8,   8,   8,   8,   8,  32 ],
		CFG_1WKEYPADS : [   8,   8,   8,   8,   8,   8,   8,   0,   0 ],
		CFG_2WKEYPADS : [   2,   2,   2,   2,   2,   2,   2,   8,  32 ],
		CFG_SIRENS    : [   2,   2,   2,   2,   2,   2,   2,   4,   8 ],
		CFG_USERCODES : [   8,   8,   8,   8,   8,   8,   8,   8,  48 ],
		CFG_PROXTAGS  : [   0,   0,   8,   0,   8,   8,   0,   8,  32 ],
		CFG_WIRELESS  : [  28,  28,  28,  28,  28,  28,  28,  29,  62 ], // 30, 64
		CFG_WIRED     : [   2,   2,   2,   2,   2,   2,   1,   1,   2 ],
		CFG_ZONECUSTOM: [   0,   5,   5,   5,   5,   5,   5,   5,   5 ]
	},

	// PMAX EEPROM CONFIGURATION version 1_2
	panelSettings: {
		// USER SETTINGS
		usrPhoneNrs   : { count:4, name:["1st Private Tel. No.","2nd Private Tel. No.","3rd Private Tel. No.","4th Private Tel. No."],
		  type:"PHONE", size:64, poff:310, psize:64, pstep:8 }, // 310, 318, 326, 334
		usrVoice      : { name:"Set Voice Option", type:"BYTE", size:8, poff:763, psize:8, posvalues:{
		  '0':"Disable Voice", '1':"Enable Voice"} },
		usrArmOption  : { name:"Auto Arm Option", type:"BYTE", size:8, poff:280, psize:1, pbitoff:5, posvalues:{
		  '1':"Enable", '0':"Disable"} },
		usrArmTime    : { name:"Auto Arm Time", type:"TIME", size:16, poff:765, psize:16 },
		usrSquawk     : { name:"Squawk Option", type:"BYTE", size:8, poff:764, psize:8, posvalues:{
		  '0':"Disable", '1':"Low Level", '2':"Medium Level", '3':"High Level"} },
		usrTimeFormat : { name:"Time Format", type:"BYTE", size:8, poff:281, psize:1, pbitoff:1, posvalues:{
		  '0':"USA - 12H", '1':"Europe - 24H"} },
		usrDateFormat : { name:"Date Format", type:"BYTE", size:8, poff:281, psize:1, pbitoff:2, posvalues:{
		  '0':"USA MM/DD/YYYY", '1':"Europe DD/MM/YYYY"} },
		// PANEL DEFINITION
		entryDelays   : { count:2, name:["Entry Delay 1","Entry Delay 2"], type:"BYTE", size:8, poff:257, psize:8, pstep:1,
		  posvalues:{'0':"None", '15':"15 Seconds", '30':"30 Seconds", '45':"45 Seconds",
		  '60':"1 Minute", '180':"3 Minutes", '240':"4 Minutes"} }, // 257, 258
		exitDelay     : { name:"Exit Delay", type:"BYTE", size:8, poff:259, psize:8, posvalues:{
		  '30':"30 Seconds", '60':"60 Seconds", '90':"90 Seconds", '120':"2 Minutes", '180':"3 Minutes", '240':"4 Minutes"} },
		bellTime      : { name:"Bell Time", type:"BYTE", size:8, poff:260, psize:8, posvalues:{
		  '1':"1 Minute", '3':"3 Minutes", '4':"4 Minutes", '8':"8 Minutes", '10':"10 Minutes", '15':"15 Minutes", '20':"20 Minutes"} },
		abortTime     : { name:"Abort Time", type:"BYTE", size:8, poff:267, psize:8, posvalues:{
		  '0':"None", '15':"15 Seconds", '30':"30 Seconds", '45':"45 Seconds",
		  '60':"1 Minute", '120':"2 Minutes", '180':"3 Minutes", '240':"4 Minutes"} },
		cancelTime    : { name:"Alarm Cancel Time", type:"BYTE", size:8, poff:266, psize:8, posvalues:{
		  '1':"1 Minute", '5':"5 Minutes", '15':"15 Minutes", '60':"60 Minutes", '240':"4 Hours", '0':"Inactive"} },
		quickArm      : { name:"Quick Arm", type:"BYTE", size:8, poff:283, psize:1, pbitoff:3, posvalues:{
		  '1':"On", '0':"Off"} },
		bypass        : { name:"Bypass", type:"BYTE", size:8, poff:284, psize:2, pbitoff:6, posvalues:{
		  '2':"Manual Bypass", '0':"No Bypass", '1':"Force Arm"} },
		exitMode      : { name:"Exit Mode", type:"BYTE", size:8, poff:282, psize:2, pbitoff:6, posvalues:{
		  '1':"Restart Exit", '2':"Off by Door", '0':"Normal"} },
		piezoBeeps    : { name:"Piezo Beeps", type:"BYTE", size:8, poff:261, psize:8, posvalues:{
		  '2':"Enable", '1':"Off when Home", '0':"Disable"} },
		troubleBeeps  : { name:"Trouble Beeps", type:"BYTE", size:8, poff:284, psize:2, pbitoff:1, posvalues:{
		  '3':"Enable", '1':"Off at Night", '0':"Disable"} },
		panicAlarm    : { name:"Panic Alarm", type:"BYTE", size:8, poff:282, psize:2, pbitoff:4, posvalues:{
		  '1':"Silent Panic", '2':"Audible Panic", '0':"Disable Panic"} },
		swingerStop   : { name:"Swinger Stop", type:"BYTE", size:8, poff:262, psize:8, posvalues:{
		  '1':"After 1 Time", '2':"After 2 Times", '3':"After 3 Times", '0':"No Shutdown"} },
		crossZoning   : { name:"Cross Zoning", type:"BYTE", size:8, poff:284, psize:1, posvalues:{
		  '1':"On", '0':"Off"} },
		supervision   : { name:"Supevision Interval", type:"BYTE", size:8, poff:264, psize:8, posvalues:{
		  '1':"1 Hour", '2':"2 Hours", '4':"4 Hours", '8':"8 Hours", '12':"12 Hours", '0':"Disable"} },
		notReady      : { name:"Not Ready", type:"BYTE", size:8, poff:281, psize:1, pbitoff:4, posvalues:{
		  '0':"Normal", '1':"In Supervision"} },
		fobAux        : { count:2, name:["Auxiliary Keyfob Button function 1","Auxiliary Keyfob Button function 2"],
		  type:"BYTE", size:"8", poff:263, psize:8, pstep:14, posvalues:{ // 263, 277
		  '1':"System Status", '2':"Instant Arm", '3':"Cancel Exit Delay", '4':"PGM/X-10"} },
		jamDetect     : { name:"Jamming Detection", type:"BYTE", size:8, poff:256, psize:8, posvalues:{
		  '1':"UL 20/20", '2':"EN 30/60", '3':"Class 6", '4':"Other", '0':"Disable"} },
		latchKey      : { name:"Latchkey Arming", type:"BYTE", size:8, poff:283, psize:1, pbitoff:7, posvalues:{
		  '1':"On", '0':"Off"} },
		noActivity    : { name:"No Activity Time", type:"BYTE", size:8, poff:265, psize:8, posvalues:{
		  '3':"3 Hours", '6':"6 Hours", '12':"12 Hours", '24':"24 Hours", '48':"48 Hours",
		  '72':"72 Hours", '0':"Disable"} },
		backLight     : { name:"Back Light Time", type:"BYTE", size:8, poff:283, psize:1, pbitoff:5, posvalues:{
		  '1':"Allways On", '0':"Off After 10 Seconds"} },
		duress        : { name:"Duress", type:"CODE", size:16, poff:273, psize:16 },
		piezoSiren    : { name:"Piezo Siren", type:"BYTE", size:8, poff:284, psize:1, pbitoff:5, posvalues:{
		  '1':"On", '0':"Off"} },
		resetOption   : { name:"Reset Option", type:"BYTE", size:8, poff:270, psize:8, posvalues:{
		  '1':"Engineer Reset", '0':"User Reset"} },
		tamperOption  : { name:"Tamper Option", type:"BYTE", size:8, poff:280, psize:1, pbitoff:1, posvalues:{
		  '1':"On", '0':"Off"} },
		sirenOnLine   : { name:"Siren On Line", type:"BYTE", size:8, poff:282, psize:1, pbitoff:1, posvalues:{
		  '1':"Enable on Fail", '0':"Disable on Fail"} },
		memoryPrompt  : { name:"Memory Prompt", type:"BYTE", size:8, poff:281, psize:1, posvalues:{
		  '1':"Enable", '0':"Disable", num:0} },
		disarmOption  : { name:"Disarm Option", type:"BYTE", size:8, poff:281, psize:2, pbitoff:6, posvalues:{
		  '0':"Any Time", '1':"On Entry All", '2':"On Entry Wireless", '3':"Entry + Away KP"} },
		bellReport    : { name:"Bell Report Option", type:"BYTE", size:8, poff:283, psize:1, posvalues:{
		  '1':"EN Standard", '0':"Others"} },
		lowBattery    : { name:"Low Battery Acknowledge", type:"BYTE", size:8, poff:281, psize:1, pbitoff:3, posvalues:{
		  '1':"On", '0':"Off"} },
		screenSaver   : { name:"Screen Saver", type:"BYTE", size:8, poff:269, psize:8, posvalues:{
		  '2':"Reset By Key", '1':"Reset By Code", '0':"Off"} },
		confirmAlarm  : { name:"Confirm Alarm Timer", type:"BYTE", size:8, poff:268, psize:8, posvalues:{
		  '0':"None", '30':"30 Minutes", '45':"45 Minutes", '60':"60 Minutes", '90':"90 Minutes"} },
		acFailure     : { name:"AC Failure Report", type:"BYTE", size:8, poff:275, psize:8, posvalues:{
		  '0':"None", '5':"5 Minutes", '30':"30 Minutes", '60':"60 Minutes", '180':"180 Minutes"} },
		userPermit    : { name:"User Permit", type:"BYTE", size:8, poff:276, psize:8, posvalues:{
		  '1':"Enable", '0':"Disable"} },
		// COMMUNICATION SETTINGS
		autotestTime  : { name:"Autotest Time", type:"TIME", size:16, poff:367, psize:16 },
		autotestCycle : { name:"Autotest Cycle", type:"BYTE", size:8, poff:369, psize:8, posvalues:{
		  '1':"1 Day", '4':"5 Days", '2':"7 Days", '3':"30 Days", '0':"Disable"} },
		areaCode      : { name:"Area Code", type:"CODE", size:24, poff:371, psize:24 },
		outAccessNr   : { name:"Out Access Number", type:"CODE", size:8, poff:374, psize:8 },
		centralStation: { count:2, name:["1st Central Station (CNTR) Tel. No.", "2nd Central Station (CNTR) Tel. No."], type:"PHONE",
		  size:64, poff:288, psize:64, pstep:11 }, // 288, 299
		accountNo     : { count:2, name:["1st Account No","2nd Account No"], type:"ACCOUNT", size:24, poff:296, psize:24, pstep:11 }, // 296, 307
		reportFormat  : { name:"Report Format", type:"BYTE", size:8, poff:363, psize:8, posvalues:{
		  '0':"Contact ID", '1':"SIA", '2':"4/2 1900/1400", '3':"4/2 1800/2300", '4':"Scancom"} },
		pulseRate     : { name:"4/2 Pulse Rate", type:"BYTE", size:8, poff:364, psize:8, posvalues:{
		  '0':"10 pps", '1':"20 pps", '2':"33 pps", '3':"40 pps"} },
		reportCentral : { name:"Report to Central Station", type:"BYTE", size:8, poff:359, psize:8, posvalues:{
		  '15':"All * Backup", '7':"All but Open/Close * Backup", '255':"All * All",
		  '119':"All but Open/Close * All but Open/Close", '135':"All but Alert * Alert",
		  '45':"Alarms * All but Alarms", '0':"Disable"} },
		reportConfirm : { name:"Report Confirmed Alarm", type:"BYTE", size:8, poff:285, psize:2, pbitoff:6, posvalues:{
		  '0':"Disable Report", '1':"Enable Report", '2':"Enable + Bypass"} },
		send2wv       : { name:"Send 2wv Code", type:"BYTE", size:8, poff:280, psize:1, pbitoff:6, posvalues:{
		  '1':"Send", '0':"Don't Send"} },
		voice2Central : { name:"Two-Way Voice To Central Stations", type:"BYTE", size:8, poff:366, psize:8, posvalues:{
		  '10':"Time-out 10 Seconds", '45':"Time-out 45 Seconds", '60':"Time-out 60 Seconds",
		  '90':"Time-out 90 Seconds", '120':"Time-out 2 Minutes", '1':"Ring Back", '0':"Disable"} },
		ringbackTime  : { name:"Ringback Time", type:"BYTE", size:8, poff:358, psize:8, posvalues:{
		  '1':"1 Minute", '3':"3 Minutes", '5':"5 Minutes", '10':"10 Minutes"} },
		csDialAttempt : { name:"Central Station Dialing Attempts", type:"BYTE", size:8, poff:362, psize:8, posvalues:{
		  '2':"2", '4':"4", '8':"8", '12':"12", '16':"16"} },
		ringbackNrs   : { count:4, name:["1st Ringback Tel No","2nd Ringback Tel No","3rd Ringback Tel No","4th Ringback Tel No"],
		  type:"PHONE", size:64, poff:310, psize:64, pstep:8}, // 310, 318, 326, 334,
		voice2Private : { name:"Two-Way Voice - Private Phones", type:"BYTE", size:8, poff:283, psize:1, pbitoff:6, posvalues:{
		  '0':"Disable", '1':"Enable"} },
		privateAttempt: { name:"Private Telephone Dialing Attempts", type:"BYTE", size:8, poff:365, psize:8, posvalues:{
		  '1':"1 Attempt", '2':"2 Attempts", '3':"3 Attempts", '4':"4 Attempts"} },
		privateReport : { name:"Reporting To Private Tel", type:"BYTE", size:8, poff:361, psize:8, posvalues:{
		  '15':"All", '7':"All but Open/Close", '13':"All but Alerts", '1':"Alarms",
		  '2':"Alerts", '8':"Open/Close", '0':"Disable Report"} },
		privateAck    : { name:"Private Telephone Acknowledge", type:"BYTE", size:8, poff:285, psize:1, pbitoff:1, posvalues:{
		  '0':"Single Acknowledge", '1':"All Acknowledge"} },
		pagerNr       : { name:"Pager Tel Number", type:"PHONE", size:64, poff:342, psize:64 },
		pagerPIN      : { name:"Pager PIN #", type:"PHONE", size:64, poff:350, psize:64 },
		pagerReport   : { name:"Report To Pager", type:"BYTE", size:8, poff:360, psize:8, posvalues:{
		  '15':"All", '3':"All + Alerts", '7':"All but Open/Close", '12':"Troubles+Open/Close",
		  '4':"Troubles", '8':"Open/Close", '0':"Disable Report"} },
		recentClose   : { name:"Recent Close Report", type:"BYTE", size:8, poff:0x11C, psize:1, pbitoff:3, posvalues:{
		  '1':"On", '0':"Off"} },
		remoteAccess  : { name:"Remote Access", type:"BYTE", size:8, poff:0x11D, psize:1, pbitoff:2, posvalues:{
		  '1':"On", '0':"Off"} },
		installerCode : { name:"Installer Code", type:"CODE", size:0x10, poff:0x20A, psize:0x10 },
		masterCode    : { name:"Master Code", type:"CODE", size:0x10, poff:0x20C, psize:0x10 },
		masterDlCode  : { name:"Master Download Code", type:"CODE", size:0x10, poff:0x20E, psize:0x10 },
		instalDlCode  : { name:"Installer Download Code", type:"CODE", size:0x10, poff:0x210, psize:0x10 },
		zoneRestore   : { name:"Zone Restore", type:"BYTE", size:8, poff:0x118, psize:1, posvalues:{
		  '0':"Report Restore", '1':"Don't Report"} },
		uploadOption  : { name:"Upload Option", type:"BYTE", size:8, poff:0x11A, psize:1, pbitoff:2, posvalues:{
		  '0':"When System Off", '1':"Any Time"} },
		dialMethod    : { name:"Dialing Method", type:"BYTE", size:8, poff:0x11D, psize:1, posvalues:{
		  '0':"Tone (DTMF)", '1':"Pulse"} },
		lineFailure   : { name:"Line Failure Report", type:"BYTE", size:8, poff:375, psize:8, posvalues:{
		  '0':"Don't Report", '1':"Immediately", '5':"5 Minutes", '30':"30 Minutes",
		  '60':"60 Minutes", '180':"180 Minutes"} },
		remoteProgNr  : { name:"Remote Programmer Tel. No.", type:"PHONE", size:64, poff:376, psize:64 },
		inactiveReport: { name:"System Inactive Report", type:"BYTE", size:8, poff:384, psize:8, posvalues:{
		  '0':"Disable", '180':"7 Days", '14':"14 Days", '30':"30 Days", '90':"90 Days"} },
		ambientLevel  : { name:"Ambient Level", type:"BYTE", size:8, poff:388, psize:8, posvalues:{
		  '0':"High Level", '1':"Low Level"} },
		// GSM DEFINITIONS
		gsmInstall    : { name:"GSM Install", type:"BYTE", size:8, poff:395, psize:8, posvalues:{
		  '1':"Installed", '0':"Not Installed"} },
		gsmSmsNrs     : { count:4, name:["1st SMS Number","2nd SMS Number","3rd SMS Number","4th SMS Number"],
		  type:"PHONE", size:64, poff:396, psize:64, pstep:8 }, // 396,404,412,420
		gsmSmsReport  : { name:"Report to SMS", type:"BYTE", size:8, poff:393, psize:8, posvalues:{
		  '15':"All", '7':"All but Open/Close", '13':"All but Alerts", '1':"Alarms",
		  '2':"Alerts", '8':"Open/Close", '0':"Disable Report"} },
		gsmFailure    : { name:"GSM Line Failure", type:"BYTE", size:8, poff:394, psize:8, posvalues:{
		  '0':"Don't Report", '2':"2 Minutes", '5':"5 Minutes", '15':"15 Minutes", '30':"30 Minutes"} },
		gsmPurpose    : { name:"GSM Line Purpose", type:"BYTE", size:8, poff:392, psize:8, posvalues:{
		  '1':"GSM is Backup", '2':"GSM is Primary", '3':"GSM Only", '0':"SMS Only"} },
		gsmAntenna    : { name:"Select GSM antenna", type:"BYTE", size:8, poff:447, psize:8, posvalues:{
		  '0':"Internal antenna", '1':"External antenna", '2':"Auto detect"} },
		// DEFINE POWERLINK
		plFailure     : { name:"PowerLink Failure", type:"BYTE", size:8, poff:391, psize:8, posvalues:{
		  '1':"Report", '0':"Disable Report"} },
		// PGM DEFINITION
		pgmPulseTime  : { name:"Pulse Time", type:"BYTE", size:8, poff:681, psize:8, posvalues:{
		  '2':"2 Seconds", '30':"30 Seconds", '120':"2 Minutes", '240':"4 Minutes"} },
		pgmByArmAway  : { name:"By Arm Away", type:"BYTE", size:8, poff:537, psize:8, posvalues:{
		  '0':"Disable", '1':"Turn Off", '2':"Turn On", '3':"Pulse Active"} },
		pgmByArmHome  : { name:"By Arm Home", type:"BYTE", size:8, poff:553, psize:8, posvalues:{
		  '0':"Disable", '1':"Turn Off", '2':"Turn On", '3':"Pulse Active"} },
		pgmByDisarm   : { name:"By Disarm", type:"BYTE", size:8, poff:569, psize:8, posvalues:{
		  '0':"Disable", '1':"Turn Off", '2':"Turn On", '3':"Pulse Active"} },
		pgmByMemory   : { name:"By Memory", type:"BYTE", size:8, poff:601, psize:8, posvalues:{
		  '0':"Disable", '1':"Turn Off", '2':"Turn On", '3':"Pulse Active"} },
		pgmByDelay    : { name:"By Delay", type:"BYTE", size:8, poff:585, psize:8, posvalues:{
		  '0':"Disable", '1':"Turn Off", '2':"Turn On", '3':"Pulse Active"} },
		pgmByKeyfob   : { name:"By Keyfob", type:"BYTE", size:8, poff:617, psize:8, posvalues:{
		  '0':"Disable", '1':"Turn Off", '2':"Turn On", '3':"Pulse Active", '4':"Toggle"} },
		pgmByLineFail : { name:"By Line Fail", type:"BYTE", size:8, poff:280, psize:1, pbitoff:2, posvalues:{
		  '1':"Yes", '0':"No"} },
		pgmZone       : { count:3, name:["Zone A","Zone B","Zone C"], type:"CODE", size:8, poff:697, psize:8, pstep:1 }, // 697, 698, 699
		pgmActZone    : { count:3, name:["Act Zone A","Act Zone B","Act Zone C"], type:"BYTE", size:8, poff:633, psize:8, pstep:16,
		  posvalues:{'0':"Disable", '1':"Turn Off", '2':"Turn On", '3':"Pulse Active"} }, // 633, 649, 665
		// DEFINE INTERNAL
		intStrobe     : { name:"Internal/Strobe Siren", type:"BYTE", size:8, poff:283, psize:1, pbitoff:1, posvalues:{
		  '0':"Internal Siren", '1':"Strobe"} },
		// X-10 GENERAL DEFINITION
		x10HouseCode  : { name:"House Code", type:"BYTE", size:8, poff:536, psize:8, posvalues:{
		  '0':"A", '1':"B", '2':"C", '3':"D", '4':"E", '5':"F", '6':"G", '7':"H", '8':"I", '9':"J", '10':"K",
		  '11':"L", '12':"M", '13':"N", '14':"O", '15':"P"} },
		x10Flash      : { name:"Flash On Alarm", type:"BYTE", size:8, poff:281, psize:1, pbitoff:5, posvalues:{
		  '1':"All Lights Flash", '0':"No Flash"} },
		x10Trouble    : { name:"Trouble Indication", type:"BYTE", size:8, poff:747, psize:8, posvalues:{
		  '1':"Enable", '0':"Disable"} },
		x10ReportCs1  : { name:"Report on Fail to Central Station 1", type:"BYTE", size:8, poff:749, psize:1,
		  posvalues:{'1':"Enable", '0':"Disable"} },
		x10ReportCs2  : { name:"Report on Fail to Central Station 2", type:"BYTE", size:8, poff:749, psize:1,
		  pbitoff:1, posvalues:{'1':"Enable", '0':"Disable"} },
		x10ReportPagr : { name:"Report on Fail to Pager", type:"BYTE", size:8, poff:749, psize:1,
		  pbitoff:2, posvalues:{'1':"Enable", '0':"Disable"} },
		x10ReportPriv : { name:"Report on Fail to Private", type:"BYTE", size:8, poff:749, psize:1,
		  pbitoff:3, posvalues:{'1':"Enable", '0':"Disable"} },
		x10ReportSMS : { name:"Report on Fail to SMS", type:"BYTE", size:8, poff:749, psize:1,
		  pbitoff:4, posvalues:{'1':"Enable", '0':"Disable"} },
		x10Lockout   : { name:"Lockout Time (start HH:MM)", type:"TIME", size:16, poff:532, psize:16 },
		x10Phase     : { name:"3 Phase and frequency", type:"BYTE", size:8, poff:748, psize:8, posvalues:{
		  '0':"Disable", '1':"50 Hz", '2':"60 Hz"} },
		// ZONES
		zoneNameRaw  : { count:31, name:"Zone name <x>", type:"STRING", size:0x80, poff:0x1900, psize:0x80, pstep:0x10 }
	},

	panelModel: {
		'0000': "PowerMax",
		'0001': "PowerMax LT",
		'0004': "PowerMax A",
		'0005': "PowerMax",
		'0006': "PowerMax LT",
		'0009': "PowerMax B",
		'0009': "PowerMax B",
		'000a': "PowerMax A",
		'000b': "PowerMax",
		'000c': "PowerMax LT",
		'000f': "PowerMax B",
		'0014': "PowerMax A",
		'0015': "PowerMax",
		'0016': "PowerMax",
		'0017': "PowerArt",
		'0018': "PowerMax SC",
		'0019': "PowerMax SK",
		'001a': "PowerMax SV",
		'001b': "PowerMax T",
		'001e': "PowerMax WSS",
		'001f': "PowerMax Smith",
		'0100': "PowerMax+",
		'0103': "PowerMax+ UK (3)",
		'0104': "PowerMax+ JP",
		'0106': "PowerMax+ CTA",
		'0108': "PowerMax+",
		'010a': "PowerMax+ SH",
		'010b': "PowerMax+ CF",
		'0112': "PowerMax+ WSS",
		'0113': "PowerMax+ 2INST",
		'0114': "PowerMax+ HL",
		'0115': "PowerMax+ UK",
		'0116': "PowerMax+ 2INST3",
		'0118': "PowerMax+ CF",
		'0119': "PowerMax+ 2INST",
		'011a': "PowerMax+",
		'011c': "PowerMax+ WSS",
		'011d': "PowerMax+ UK",
		'0120': "PowerMax+",
		'0121': "PowerMax+",
		'0122': "PowerMax+ CF",
		'0124': "PowerMax+ UK",
		'0127': "PowerMax+ 2INST_MONITOR",
		'0128': "PowerMax+ KeyOnOff",
		'0129': "PowerMax+ 2INST_MONITOR",
		'012a': "PowerMax+ 2INST_MONITOR42",
		'012b': "PowerMax+ 2INST33",
		'012c': "PowerMax+ One Inst_1_44_0",
		'012d': "PowerMax+ CF_1_45_0",
		'012e': "PowerMax+ SA_1_46",
		'012f': "PowerMax+ UK_1_47/UK_3_4_17",
		'0130': "PowerMax+ SA UK_1_48",
		'0132': "PowerMax+ KeyOnOff 1_50",
		'0201': "PowerMax Pro",
		'0202': "PowerMax Pro-Nuon ",
		'0204': "PowerMax Pro-PortugalTelecom",
		'020a': "PowerMax Pro-PortugalTelecom2",
		'020c': "PowerMax HW-V9 Pro",
		'020d': "PowerMax ProSms",
		'0214': "PowerMax Pro-PortugalTelecom_4_5_02",
		'0216': "PowerMax HW-V9_4_5_02 Pro",
		'0217': "PowerMax ProSms_4_5_02",
		'0218': "PowerMax UK_DD243_4_5_02 Pro M",
		'021b': "PowerMax Pro-Part2__2_27",
		'0223': "PowerMax Pro Bell-Canada",
		'0301': "PowerMax Complete",
		'0302': "PowerMax Complete_NV",
		'0303': "PowerMax Complete-PortugalTelecom",
		'0307': "PowerMax Complete_1_0_07",
		'0308': "PowerMax Complete_NV_1_0_07",
		'030a': "PowerMax Complete_UK_DD243_1_1_03",
		'030b': "PowerMax Complete_COUNTERFORCE_1_0_06",
		'0401': "PowerMax Pro-Part",
		'0402': "PowerMax Pro-Part CellAdaptor",
		'0405': "PowerMax Pro-Part_5_0_08",
		'0406': "PowerMax Pro-Part CellAdaptor_5_2_04",
		'0407': "PowerMax Pro-Part KeyOnOff_5_0_08",
		'0408': "PowerMax UK Pro-Part_5_0_08",
		'0409': "PowerMax SectorUK Pro-Part_5_0_08",
		'040a': "PowerMax Pro-Part CP1 4_10",
		'040c': "PowerMax Pro-Part_Cell_key_4_12",
		'040d': "PowerMax Pro-Part UK 4_13",
		'040e': "PowerMax SectorUK Pro-Part_4_14",
		'040f': "PowerMax Pro-Part UK 4_15",
		'0410': "PowerMax Pro-Part CP1 4_16",
		'0411': "PowerMax NUON key 4_17",
		'0433': "PowerMax Pro-Part2__4_51",
		'0434': "PowerMax UK Pro-Part2__4_52",
		'0436': "PowerMax Pro-Part2__4_54",
		'0437': "PowerMax Pro-Part2__4_55 (CP_01)",
		'0438': "PowerMax Pro-Part2__4_56",
		'0439': "PowerMax Pro-Part2__4_57 (NUON)",
		'043a': "PowerMax Pro 4_58",
		'043c': "PowerMax Pro 4_60",
		'043e': "PowerMax Pro-Part2__4_62",
		'0440': "PowerMax Pro-Part2__4_64",
		'0442': "PowerMax 4_66",
		'0443': "PowerMax Pro 4_67",
		'0444': "PowerMax Pro 4_68",
		'0445': "PowerMax Pro 4_69",
		'0446': "PowerMax Pro-Part2__4_70",
		'0447': "PowerMax 4_71",
		'0449': "PowerMax 4_73",
		'044b': "PowerMax Pro-Part2__4_75",
		'0451': "PowerMax Pro 4_81",
		'0452': "PowerMax Pro 4_82",
		'0454': "PowerMax 4_84",
		'0455': "PowerMax 4_85",
		'0456': "PowerMax 4_86",
		'0457': "PowerMax 4_87",
		'0459': "PowerMax 4_89",
		'045a': "PowerMax 4_90",
		'045b': "PowerMax 4_91",
		'045c': "PowerMax 4_92",
		'0503': "PowerMax UK Complete partition 1_5_00",
		'050a': "PowerMax Complete partition GPRS",
		'050b': "PowerMax Complete partition NV GPRS",
		'050c': "PowerMax Complete partition GPRS NO-BBA",
		'050d': "PowerMax Complete partition NV GPRS NO-BBA",
		'050e': "PowerMax Complete part. GPRS NO-BBA UK_5_14",
		'0511': "PowerMax Pro-Part CP1 GPRS 5_17",
		'0512': "PowerMax Complete part. BBA UK_5_18",
		'0533': "PowerMax Complete 5_51",
		'0534': "PowerMax Complete part2 5_52 (UK)",
		'0536': "PowerMax Complete 5_54 (GR)",
		'0537': "PowerMax Complete  5_55",
		'053a': "PowerMax Complete 5_58 (PT)",
		'053b': "PowerMax Complete NV 5_59 (NV)",
		'053c': "PowerMax Complete  5_60",
		'053e': "PowerMax Complete 5_62",
		'053f': "PowerMax Complete  5_63",
		'0540': "PowerMax Complete  5_64",
		'0541': "PowerMax Complete  5_65",
		'0543': "PowerMax Complete  5_67",
		'0544': "PowerMax Complete  5_68",
		'0545': "PowerMax Complete  5_69",
		'0546': "PowerMax Complete  5_70",
		'0547': "PowerMax Complete  5_71",
		'0549': "PowerMax Complete  5_73",
		'054b': "PowerMax Complete  5_75",
		'054c': "PowerMax Complete  5_76",
		'054f': "PowerMax Complete  5_79",
		'0550': "PowerMax Complete  5_80",
		'0551': "PowerMax Complete  5_81",
		'0553': "PowerMax Complete  5_83",
		'0554': "PowerMax Complete  5_84",
		'0555': "PowerMax Complete  5_85",
		'0556': "PowerMax Complete  5_86",
		'0601': "PowerMax Express",
		'0603': "PowerMax Express CP 01",
		'0605': "PowerMax Express OEM 6_5",
		'0607': "PowerMax Express BBA 6_7",
		'0608': "PowerMax Express CP 01 BBA 6_8",
		'0609': "PowerMax Express OEM1 BBA 6_9",
		'060b': "PowerMax Express BBA 6_11",
		'0633': "PowerMax Express 6_51",
		'063b': "PowerMax Express 6_59",
		'063d': "PowerMax Express 6_61",
		'063e': "PowerMax Express 6_62 (UK)",
		'0645': "PowerMax Express 6_69",
		'0647': "PowerMax Express 6_71",
		'0648': "PowerMax Express 6_72",
		'0649': "PowerMax Express 6_73",
		'064a': "PowerMax Activa 6_74",
		'064c': "PowerMax Express 6_76",
		'064d': "PowerMax Express 6_77",
		'064e': "PowerMax Express 6_78",
		'064f': "PowerMax Express 6_79",
		'0650': "PowerMax Express 6_80",
		'0651': "PowerMax Express 6_81",
		'0652': "PowerMax Express 6_82",
		'0653': "PowerMax Express 6_83",
		'0654': "PowerMax  6_84",
		'0655': "PowerMax  6_85",
		'0658': "PowerMax  6_88",
		'0659': "PowerMax  6_89",
		'065a': "PowerMax  6_90",
		'065b': "PowerMax  6_91",
		'065c': "PowerMax  6_92",
		'065d': "PowerMax  6_93",
		'065e': "PowerMax  6_94",
		'0701': "PowerMax PowerCode-G 7_1",
		'0702': "PowerMax PowerCode-G 7_2",
		'0704': "PowerMaster10 7_4",
		'0705': "PowerMaster10 7_05",
		'0707': "PowerMaster10 7_07",
		'070c': "PowerMaster10 7_12",
		'070f': "PowerMaster10 7_15",
		'0710': "PowerMaster10 7_16",
		'0711': "PowerMaster10 7_17",
		'0712': "PowerMaster10 7_18",
		'0713': "PowerMaster10 7_19",
		'0714': "PowerMaster10 7_20",
		'0716': "PowerMaster10 7_22",
		'0717': "PowerMaster10 7_23",
		'071e': "PowerMaster10 7_30",
		'071f': "PowerMaster10 7_31",
		'0720': "PowerMaster10 7_32",
		'0721': "PowerMaster10 7_33",
		'0722': "PowerMaster10 7_34",
		'0723': "PowerMaster10 7_35",
		'0724': "PowerMaster10 7_36",
		'072d': "PowerMaster10 7_45",
		'072e': "PowerMaster10 7_46",
		'0730': "PowerMaster10 7_48",
		'0731': "PowerMaster10 7_49",
		'0733': "PowerMaster10 7_51",
		'0734': "PowerMaster10 7_52",
		'0735': "PowerMaster10 7_53",
		'0736': "PowerMaster10 7_54",
		'0737': "PowerMaster10 7_55",
		'0738': "PowerMaster10 7_56",
		'0739': "PowerMaster10 7_57",
		'073a': "PowerMaster10 7_58",
		'073b': "PowerMaster10 7_59",
		'073c': "PowerMaster10 7_60",
		'073e': "PowerMaster10 7_62",
		'073f': "PowerMaster10 7_63",
		'0743': "PowerMaster10 7_67",
		'0747': "PowerMaster10 7_71",
		'0748': "PowerMaster10 7_72",
		'0751': "PowerMaster10 7_81",
		'075a': "PowerMaster10 7_90",
		'0764': "PowerMaster10 7_100",
		'0765': "PowerMaster10 7_101",
		'0766': "PowerMaster10 7_102",
		'0767': "PowerMaster10 7_103",
		'0768': "PowerMaster10 7_104",
		'0773': "PowerMaster10 7_115",
		'0774': "PowerMaster10 7_116",
		'0775': "PowerMaster10 7_117",
		'0776': "PowerMaster10 7_118",
		'077e': "PowerMaster10 7_126",
		'0802': "PowerMax Complete PowerCode-G 8_2",
		'0803': "PowerMaster30 8_3",
		'080f': "PowerMaster30 8_15",
		'0810': "PowerMaster30 8_16",
		'0812': "PowerMaster30 8_18",
		'0813': "PowerMaster30 8_19",
		'0814': "PowerMaster30 8_20",
		'0815': "PowerMaster30 8_21",
		'081e': "PowerMaster30 8_30",
		'081f': "PowerMaster30 8_31",
		'0820': "PowerMaster30 8_32",
		'0821': "PowerMaster30 8_33",
		'0823': "PowerMaster30 8_35",
		'0824': "PowerMaster30 8_36",
		'082d': "PowerMaster30 8_45",
		'082e': "PowerMaster30 8_46",
		'0830': "PowerMaster30 8_48",
		'0831': "PowerMaster30 8_49",
		'0832': "PowerMaster30 8_50",
		'0833': "PowerMaster30 8_51",
		'0834': "PowerMaster30 8_52",
		'0835': "PowerMaster30 8_53",
		'0836': "PowerMaster30 8_54",
		'0838': "PowerMaster30 8_56",
		'0839': "PowerMaster30 8_57",
		'083a': "PowerMaster30 8_58",
		'083c': "PowerMaster30 8_60",
		'083e': "PowerMaster30 8_62",
		'083f': "PowerMaster30 8_63",
		'0843': "PowerMaster30 8_67",
		'0847': "PowerMaster30 8_71",
		'0848': "PowerMaster30 8_72",
		'0849': "PowerMaster30 8_73",
		'0851': "PowerMaster30 8_81",
		'085a': "PowerMaster30 8_90",
		'085c': "PowerMaster30 8_92",
		'0864': "PowerMaster30 8_100",
		'0865': "PowerMaster30 8_101",
		'0873': "PowerMaster30 8_115",
		'0874': "PowerMaster30 8_116",
		'0875': "PowerMaster30 8_117",
		'087d': "PowerMaster30 8_125",
		'0a3f': "PowerMaster3310_63",
		'0a40': "PowerMaster3310_64",
		'0a43': "PowerMaster33 10_67",
		'0a44': "PowerMaster3310_68",
		'0a47': "PowerMaster3310_71",
		'0a48': "PowerMaster33 10_72",
		'0a51': "PowerMaster3310_81",
		'0a64': "PowerMaster33 10_100",
		'0a73': "PowerMaster33 10_115",
		'0a74': "PowerMaster3310_116",
		'0d50': "MSP5 13_80",
		'0d55': "PM360 13_85",
		'0d57': "PM360 13_87",
		'0d58': "PM360 13 88",
		'0d5b': "PM360 13 91",
		'0d64': "PM360 13 100",
		'0d65': "PM360 13_101",
		'0d66': "PM360 13 102",
		'0d73': "PM360 13 115",
		'0d7e': "PM360 13 126",
		'0f88': "PowerMaster30 8_116",
		'0f8d': "PM33EXP 15_141"
	},

	// "Arming Key", "Guard" ??
	zoneType: {
		en: [
			"Non-Alarm", "Emergency", "Flood", "Gas", "Delay 1", "Delay 2", "Interior-Follow", "Perimeter", "Perimeter-Follow",
			"24 Hours Silent", "24 Hours Audible", "Fire", "Interior", "Home Delay", "Temperature", "Outdoor", "16"
		],
		nl: [
			"Geen alarm", "Noodtoestand", "Water", "Gas", "Vertraagd 1", "Vertraagd 2", "Interieur volg", "Omtrek", "Omtrek volg",
			"24 uurs stil", "24 uurs luid", "Brand", "Interieur", "Thuis vertraagd", "Temperatuur", "Buiten", "16"
		]
	},

	// Zone names are taken from the panel, so no language support needed
	zoneName: [
		"Attic", "Back door", "Basement", "Bathroom", "Bedroom", "Child room", "Closet", "Den", "Dining room", "Downstairs",
		"Emergency", "Fire", "Front door", "Garage", "Garage door", "Guest room", "Hall", "Kitchen", "Laundry room", "Living room",
		"Master bathroom", "Master bedroom", "Office", "Upstairs", "Utility room", "Yard", "Custom 1", "Custom 2", "Custom 3",
		"Custom4", "Custom 5", "Not Installed"
	],

	zoneChime: {
		en: [ 'Off', 'Melody', 'Zone' ],
		nl: [ 'Uit', 'Muziek', 'Zone' ]
	},

	// Note: names need to match to VAR_xxx
	// unknown to date: Push Button, Flood, Universal
	zoneSensor: {
		'0x3': "Motion", '0x4': "Motion", '0x5': "Magnet", '0x6': "Magnet", '0x7': "Magnet", '0xa': "Smoke", '0xb': "Gas",
		'0xc': "Motion", '0xf': "Wired"
	},

	// PowerMaster specific
	zoneSensorMaster: {
		'0x1': { name: "Next PG2", func: "Motion" },
		'0x4': { name: "Next CAM PG2", func: "Camera" },
		'0x7': { name: "SD-304 PG2", func: "Magnet" },
		'0x15': { name: "SMD-427 PG2", func: "Smoke" },
		'0x16': { name: "SMD-426 PG2", func: "Smoke" },
		'0x18': { name: "GSD-442 PG2", func: "Gas" },
		'0x1a': { name: "TMD-560 PG2", func: "Temperature" },
		'0x29': { name: "MC-302V PG2", func: "Magnet" },
		'0x2a': { name: "MC-302 PG2", func: "Magnet" },
		'0x35': { name: "TOWER-32AM PG2", func: "Motion" },
		'0xfe': { name: "Wired sensor", func: "Wired" }
	},

	// Calculate CRC for PDU send to PowerMax
	calcCRC: function(pdu) {
		var checksum = 0;
		var len = pdu.length;

		if (len <= 2) { return false }
		for (var i = 1; i <= (len - 3); i++) {
			checksum += pdu[i];
		}
		checksum = 0xFF - (checksum % 0xFF);
		if (checksum == 0xFF) { checksum = 0x00; }
		var newPDU = pdu.slice(0, len - 2);
		newPDU.push(checksum, 0x0A);
		return newPDU
	},

	// Check CRC in PDU received from PowerMax
	checkCRC: function(pdu) {
		var checksum = 0;
		var len = pdu.length;

		if (len <= 2) { return false }
		for (var i = 1; i <= (len - 3); i++) {
			checksum += pdu[i];
		}
		checksum = 0xFF - (checksum % 0xFF);
		if (checksum == 0xFF) { checksum = 0x00; }
		var got = pdu[len - 2];
		if (checksum == got) {
			return true
		} else {
			//'CRC check failed (expected ' + checksum + ', got ' + got + ')';
			return false;
		}
	},

	secret: [0x56, 0x50],

	info: {
		type: { en: 'Panel type', nl: 'Type paneel', val: 'panelType' }, // e.g. 'PowerMax Pro Part'
		name: { en: 'Panel name', nl: 'Naam paneel', val: 'panelName' }, // e.g. 'PowerMax Pro-Part2__4_54'
		serial: { en: 'Serial number', nl: 'Serienummer', val: 'panelSerial' },
		software: { en: 'Software version', nl: 'Versienummer software', val: 'panelSoftware' }
	}
}

module.exports = PowerMaxTables;
